docs/src/userguide/language_basics.rst

```diff
@@ -760,12 +758,27 @@ An external C++ function that may raise an exception can be declared with::
 
 See :ref:`wrapping-cplusplus` for more details.
 
+Finally, if you are certain that your function should not raise an exception, (e.g., it
+does not use Python objects at all, or you plan to use it as a callback in C code that
+is unaware of Python exceptions), you can declare it as such using ``noexcept``::
+
+    cdef int spam() noexcept
+
+If a ``noexcept`` function *does* finish with an exception then it will print a warning message but not allow the exception to propagate further.
 Some things to note:
 
+* ``cdef`` functions that are also ``extern`` are implicitly declared ``noexcept``.
+  In the uncommon case of external C/C++ functions that _can_  raise Python exceptions,
+  e.g., external functions that use the Python C API, you should explicitly declare
+  them with an exception value.
+
+* ``cdef`` functions that are *not* ``extern`` are implicitly declared ``except *``.
```

Comment:
This isn't quite true I think.

```suggestion
* ``cdef`` functions that are *not* ``extern`` are implicitly declared with a suitable exception specification for the return type (e.g. ``except *`` for a ``void`` return type, ``except -1`` for an ``int`` return type).
```

I'm a bit unsure how much detail we should go into here so feel free to reword it if you want. But `except -1` is likely to be more efficient than `except *` so I think it's worth saying

Reply:
This is tangential to your suggestion, which I will incorporate, but is it also true that `except *` is also "optimized" to `except -1` for `int` for example?

Reply:
>  but is it also true that except * is also "optimized" to except -1 for int for example?

Thinking about it more I should have said `except? -1`. But yes, it's true for non-extern functions.